### PR TITLE
Implement `lists::concatenate_list_elements`

### LINKED ISSUE #8164: [FEA] Support concatenate_list_elements for list type
Currently, we have `strings::concatenate_list_elements` to concatenate strings within each list (row) in a lists column. We also need the same functionality for list type (in particular, column of lists of lists). For example:
```
l = [ [{1, 2}, {3, 4}, {5}], [{6}, {}, {7, 8, 9}] ]
r = lists::concatenate_list_elements(l);
r is [ {1, 2, 3, 4, 5}, {6, 7, 8, 9} ] 
```

### PR DESCRIPTION
**Pull request by ttnghia**
This PR implements `lists::concatenate_list_elements` for list type. Given a lists column in which each row is a list of lists, the output column is generated by concatenating all lists in the same row into a single list.

Example:
```
l = [ [{1, 2}, {3, 4}, {5}], [{6}, {}, {7, 8, 9}] ]
r = lists::concatenate_list_elements(l);
r is [ {1, 2, 3, 4, 5}, {6, 7, 8, 9} ]
```

This closes #8164. In addition, `lists::concatenate_rows` is rewritten using `lists::interleave_columns` following by `lists::concatenate_list_elements`, which is significantly shorter.

### CODE REVIEW

**Comment by ttnghia** [comment]
Reference: https://github.com/rapidsai/cudf/pull/8006

**Review by robertmaynard — APPROVED** [review]
Approving CMake changes.

**Review by jlowe — APPROVED** [review]
Java approval

**Review by ajschmidt8 — APPROVED** [review]
Approving `ops-codeowner` file changes

**Review by devavret — CHANGES_REQUESTED** [review]
This code is giving me major déjà vu. Can you try to consolidate this with `lists::concatinate_rows`? There seems to be lot of shared logic. Even better, it looks like the entirety of `lists::concatinate_rows` can be reimplemented as an `interleave_columns` followed by a `concatenate_list_elements`.

File: cpp/src/lists/combine/concatenate_list_elements.cu [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+  auto type = input.type();  // Column that is lists of lists.
+  CUDF_EXPECTS(type.id() == type_id::LIST, "Input column must be a lists column.");
+
+  auto col = lists_column_view(input).child();  // Rows, which are lists.
+  type     = col.type();
+  CUDF_EXPECTS(type.id() == type_id::LIST, "Rows of the input column must be lists.");
+
+  col  = lists_column_view(col).child();  // Lowest level entries, which are of non-nested types.
+  type = col.type();
+  CUDF_EXPECTS(not cudf::is_nested(type),
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/src/lists/combine/concatenate_list_elements.cu:377]: Good idea. I will look into consolidate `lists::concatenate_rows`. Thanks.

For many level nested lists column, concatenate the lists will require recursive call of list concatenation on the child columns. This is very inefficient when implemented in device code.

[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/src/lists/combine/concatenate_list_elements.cu]: Done. Please have a look.

File: python/cudf/cudf/_lib/cpp/lists/concatenate_rows.pxd [hunk 2]
```diff
@@ -5,7 +5,7 @@ from libcpp.memory cimport unique_ptr
 from cudf._lib.cpp.column.column cimport column
 from cudf._lib.cpp.table.table_view cimport table_view
 
-cdef extern from "cudf/lists/concatenate_rows.hpp" namespace \
+cdef extern from "cudf/lists/combine.hpp" namespace \
```
Comment (galipremsagar) [inline python/cudf/cudf/_lib/cpp/lists/concatenate_rows.pxd:8]: This file would likely have to be renamed too to `combine.pxd`

**Review by galipremsagar — CHANGES_REQUESTED** [review]

File: python/cudf/cudf/_lib/cpp/lists/concatenate_rows.pxd (hunk 2 above)
Comment (ttnghia) [reply python/cudf/cudf/_lib/cpp/lists/concatenate_rows.pxd:8]: Thanks. Done.

**Review by galipremsagar — APPROVED** [review]

**Review by devavret — APPROVED** [review]
This much better

**Comment by ttnghia** [comment]
Rerun tests.

**Comment by ttnghia** [comment]
@gpucibot merge

**Pull request merged**